File: frontend/src/types/annotation.ts
export type AnnotationType = 'circle' | 'box' | 'rectangle' | 'polygon';
export type AnnotationSource = 'manual' | 'simpleblob' | 'yolo' | 'sam2';

export interface CircleData {
  x: number;
  y: number;
  size: number;
}

export interface BoxData {
  corners: [number, number][];  // Array of [x, y] points
}

export interface PolygonData {
  points: [number, number][];  // Array of [x, y] points
}

export type AnnotationData = CircleData | BoxData | PolygonData;

export interface Annotation {
  id: string;
  image_id: string;
  type: AnnotationType;
  data: any;  // Will be one of the specific data types
  confidence?: number;
  source: AnnotationSource;
  class_label?: string;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface AnnotationCreate {
  type: AnnotationType;
  data: any;
  confidence?: number;
  source?: AnnotationSource;
  class_label?: string;
}

export interface AnnotationUpdate {
  data?: any;
  class_label?: string;
  confidence?: number;
}

File: backend/app/schemas/annotation.py
"""Annotation schemas"""
from pydantic import BaseModel, Field, field_validator
from uuid import UUID
from datetime import datetime
from typing import Optional, Dict, Any, Literal


class AnnotationCreate(BaseModel):
    """Schema for creating an annotation"""
    type: Literal["circle", "box", "rectangle", "polygon"]
    data: Dict[str, Any]
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    source: Literal["manual", "simpleblob", "yolo", "sam2"] = "manual"
    class_label: Optional[str] = None

    @field_validator("data")
    @classmethod
    def validate_data(cls, v, info):
        """Validate annotation data based on type"""
        annotation_type = info.data.get("type")

        if annotation_type == "circle":
            required = ["x", "y", "size"]
            if not all(k in v for k in required):
                raise ValueError(f"Circle requires: {required}")
        elif annotation_type in ["box", "rectangle"]:
            if "corners" not in v or len(v["corners"]) != 4:
                raise ValueError("Box/rectangle requires 4 corners")
        elif annotation_type == "polygon":
            if "points" not in v or len(v["points"]) < 3:
                raise ValueError("Polygon requires at least 3 points")

        return v


class AnnotationUpdate(BaseModel):
    """Schema for updating an annotation"""
    data: Optional[Dict[str, Any]] = None
    class_label: Optional[str] = None
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)


class AnnotationResponse(BaseModel):
    """Schema for annotation response"""
    id: UUID
    image_id: UUID
    type: str
    data: Dict[str, Any]
    confidence: Optional[float]
    source: str
    class_label: Optional[str]
    created_by: Optional[UUID]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
